### ISSUE TITLE
Error "failed:[ShapeInferenceError] First input does not have rank 2"

### ISSUE THREAD

**Issue by luan1412167** [microsoft/onnxruntime#2045]
**Describe the bug**
I'm facing the error when run arcface_resnet34 converted from mxnet to onnx. My input shape (1,3,112,112). I don't know the error that be where. If you know please help me solve it. Thanks
`terminate called after throwing an instance of 'Ort::Exception'
  what():  Load model from /home/luandd/CLionProjects/untitled/mxnet_resnet34.onnx failed:[ShapeInferenceError] First input does not have rank 2`
**Urgency**
If there are particular important use cases blocked by this or strict project-related timelines, please share more information and dates. If there are no hard deadlines, please specify none.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 18.04
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: 0.5.0
- Python version: no
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**To Reproduce**
Describe steps/code to reproduce the behavior:

**Expected behavior**

**Screenshots**

**Additional context**
Add any other context about the problem here. If the issue is about a particular model, please share the model details as well to facilitate debugging.

**Comment by faxu** [comment]
@hariharans29 is this related to the the other arcface model issue? https://github.com/microsoft/onnxruntime/issues/831

**Comment by hariharans29** [comment]
Most likely not. The other issue is NaN outputs. This model does not load. Most likely it is invalid due to conversion error or there is a bug to be fixed in shape inference.

@luan- could you pl share the model ?

**Comment by luan1412167** [comment]
@hariharans29 it's [here](https://drive.google.com/file/d/1sskJOezjuMvU6f5XQvNWaFW5z1xRl-w5/view?usp=sharing). 
I used arcface_resnet34 from insightface repo and converted from mxnet to onnx with spatial=1 (modify spatial parameter in mxnet/contrib/onnx/mx2onnx/_op_translations.py). 
Then I get error `2019-10-09 00:51:21.437349791 [E:onnxruntime:, sequential_executor.cc:165 Execute] Non-zero status code returned while running PRelu node. Name:'relu0' Status Message: /home/luandd/project_company/face_rec/onnxruntime/onnxruntime/core/providers/cpu/math/element_wise_ops.h:329 void onnxruntime::BroadcastIterator::Init(int64_t, int64_t) axis == 1 || axis == largest was false. Attempting to broadcast an axis by a dimension other than 1. 64 by 112
'` . 
After I reshape input prelu layer then I got the error `failed:[ShapeInferenceError] First input does not have rank 2`. Thanks for your help.

**Comment by hariharans29** [comment]
Hi,

The model you shared cannot run as it violates ONNX op spec. Hand-computing shape till the problematic PRelu node - 

![image](https://user-images.githubusercontent.com/9969784/66422357-6475a380-e9be-11e9-9051-3e51250c732c.png)

1) The input, Sub output, Mul output have the shape [1, 3, 112, 112]

2) The output shape of Conv will be (based on the parameters) - [1, 64, 112, 112]

3) The PRelu node has slope shape [64]. The spec [here](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#prelu-7) says that the slope shape must be unidirectionally broadcastable to the input shape [64] is NOT unidirectiaonally broadcastable to [1, 64, 112, 112]. [64, 1, 1] is unidirectionally broadcastable, you can try that and see if the results look okay.

This fix is needed in all the PRelu nodes. I think this is a bug for the conversion tool that you are using to convert into ONNX. It generates an invalid ONNX model.

**Comment by luan1412167** [comment]
@hariharans29 I have reshaped prelu input to (64,4) as following image. I still facing error  `failed:[ShapeInferenceError] First input does not have rank 2` . the model after reshape [here](https://drive.google.com/file/d/1O3GYMZwBuexaxNo3Wjp9CLHhpgcVrAU9/view?usp=sharing)
![Screenshot from 2019-10-09 09-15-52](https://user-images.githubusercontent.com/37315746/66446607-7a926b00-ea75-11e9-9c5f-e614708079f1.png)

**Comment by hariharans29** [comment]
The model still seems to have some issues-

![image](https://user-images.githubusercontent.com/9969784/66447047-d0c5c680-ea01-11e9-8ce8-aacd4aaaac5c.png)

The output shape of the BatchNormalization node (above Gemm) should be 4D (rank == 4) and dropout preserves the shape and Gemm expects 2D input (rank == 2). Possibly, missing some reshape there ?

**Comment by luan1412167** [comment]
@hariharans29  thanks for your support. I have added flatten layer into it like [here](https://github.com/apache/incubator-mxnet/pull/13356/commits/647f30ec37033602ec87e4ac6142145e0806c5a9) . It worked

**Comment by hariharans29** [comment]
Thanks for confirming that it worked. Please validate that the result aligns with the original base framework model. 

Closing this issue.

**Comment by luan1412167** [comment]
@hariharans29 then I validate 2 these images. cosine distance result is [[0.9988895]]
It seem to be wrong.
![Tom_Hanks_54745](https://user-images.githubusercontent.com/37315746/66453085-53e02e80-ea8d-11e9-96cf-8eef91df885d.png)
![images](https://user-images.githubusercontent.com/37315746/66453099-5e022d00-ea8d-11e9-8673-e09107d9db0a.jpeg)
My modified model [here](https://drive.google.com/file/d/1O3GYMZwBuexaxNo3Wjp9CLHhpgcVrAU9/view?usp=sharing) 
What is the wrong @hariharans29 ?

**Comment by hariharans29** [comment]
What is the expected cosine distance ?

I don't think you can arbitrarily change spatial 0 to 1 and expect the model to work with no change. It must match exactly the original models graph, otherwise results will not match. 

I feel like we made some ad-hoc changes to the onnx graph just to make it pass in ORT and hence the results differed. 

I would suggest that you check with the mxnet onnx exporter team on what's the issue with the original exported model....

**Comment by luan1412167** [comment]
@hariharans29 I think you had much experiments with spatial when convert from mxnet to onnx
https://github.com/microsoft/onnxruntime/pull/890#issue-272928157
https://github.com/onnx/models/issues/156#issue-439851958
I would open discussion with team onnx export team. Meanwhile Can you suggest me a way can run model with correct result.

**Comment by hariharans29** [comment]
1) How much are the results off by ?

2) I am afraid I cannot help with model conversion issue as I am not familiar with the conversion logics involved. Please try and get a *valid* equivalent onnx model from the mxnet onnx export team. We can go from there.

3) Maybe the last Flatten hack was incorrect. Should it be reshaped to something else ? (just a thought - not sure it will work)

**Comment by luan1412167** [comment]
@hariharans29 Given your experiments, Have anyone run arcface model on oxxn runtime?

**Comment by hariharans29** [comment]
We *may* add support for spatial == 0 soon. You can try it out then. But you need to get a valid ONNX model with spatial == 0 to try it.

**Comment by luan1412167** [comment]
@hariharans29  I is have valid onnx model with spatial=0 however I can't run it on current onnx version. If you support for spatial=0 please ping to me. thanks

**Comment by hariharans29** [comment]
You can try after https://github.com/microsoft/onnxruntime/pull/2092 is merged. But please be advised that the MxNet exporter seems to stamp spatial == 0 to the BatchNorm nodes when it actually means spatial == 1. This was observed in the ArcFace model from the ONNX model zoo - https://github.com/onnx/models/issues/156.

If after that, the results still don't match, you will have to check if there were issues with the conversion by talking to the relevant people who exported the graph to ONNX from the base framework.

**Comment by luan1412167** [comment]
@hariharans29 I would try with #2092.However, In bad case, whether pytorch converter have get valid result?

**Comment by hariharans29** [comment]
I don't know. I haven't tried. Please give it a shot.

**Comment by hariharans29** [comment]
Most likely if the inbox model is valid after conversion, it will be fine. It is very rare if conversion succeeds but results don't match.

**Comment by luan1412167** [comment]
@hariharans29  thanks for your support. I will inform result when I try it.

**Comment by hariharans29** [comment]
Hi luan,

1) If you have immediate ArcFace inferencing needs, please check out my reply in https://github.com/onnx/models/issues/156, on how to "correct" the ArcFace ONNX model zoo model and please use it.

2)  If you have conversion issues (conversion yields invalid ONNX model or conversion yields an ONNX model whose results are not right), please check with the base framework (MxNet / PyTorch) ONNX conversion group.

As such, since this is not an ONNXRuntime issue now, closing this out. Thanks.

**Comment by luan1412167** [comment]
@hariharans29 Are you know this error? I'm using onnx runtime from source (latest). your model [here](https://drive.google.com/file/d/1aZBakhRE9dIhiGf8WI-hpS6nenuYcZm9/view?usp=sharing) 
![Screenshot from 2019-10-18 14-58-20](https://user-images.githubusercontent.com/37315746/67076651-ca7cda80-f1b7-11e9-8a0e-8399557ed1b4.png)

**Comment by sky186** [comment]
arcface correct model
    input_shape=(1,3,112,112)
    sym = './model-symbol.json'
    params = './model-0000.params'
    ###########################################
    onnx_file="./mynet.onnx"
    print ("************************")
    converted_model_path = onnx_mxnet.export_model(sym,
        params,
        [input_shape],
        np.float32,
        onnx_file,
        verbose=True  #print node information   data ,Output node is: softmax
    )
    model = onnx.load(r'mynet.onnx')
    for node in model.graph.node:
        if (node.op_type == "BatchNormalization"):
            for attr in node.attribute:
                if (attr.name == "spatial"):  #0 to 1
                    attr.i = 1  ## use to onnxruntime , not to effect output
    onnx.save(model, r'mynet.onnx')  


the onnx out is same the mxnet,but onnxruntime not pass,need to fix
###### Prelu修复，形状从【64】变成【1，64，1，1】
https://github.com/apache/incubator-mxnet/pull/13460/commits/f1a6df82a40d1d9e8be6f7c3f9f4dcfe75948bd6
###### 添加 ONNX export: Add Flatten before Gemm，添加flatten
https://github.com/apache/incubator-mxnet/pull/13356/files

**Comment by cyrusbehr** [comment]
Making the changes that @sky186 suggests allows me to perform inference with the model, though I am not sure if the inference results are correct... 
Anyone else given this a try?

**Comment by HoangTienDuc** [comment]
> @hariharans29 thanks for your support. I have added flatten layer into it like [here](https://github.com/apache/incubator-mxnet/pull/13356/commits/647f30ec37033602ec87e4ac6142145e0806c5a9) . It worked

hi @luan1412167 @hariharans29 how did you shape and flatten?
